nixos/modules/services/networking/peroxide.nix

```diff
@@ -0,0 +1,180 @@
+
+              <warning>
+                This <emphasis>must</emphasis> be a subdirectory of
+                <xref linked="opt-services.peroxide.stateDir"/>
+              </warning>
+            '';
+          };
+
+          UserPortImap = mkOption {
+            type = types.port;
+            default = 1143;
+            description = "The port on which to listen for IMAP connections.";
+          };
+
+          UserPortSmtp = mkOption {
+            type = types.port;
+            default = 1025;
+            description = "The port on which to listen for SMTP connections.";
+          };
+
+          ServerAddress = mkOption {
+            type = types.str;
+            default = "[::0]";
+            example = "localhost";
+            description = "The address on which to listen for connections.";
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for peroxide.  See
+        <link xlink:href="https://github.com/ljanyst/peroxide/blob/v${pkgs.peroxide.version}/config.example.yaml"/>
+        for an example configuration.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.tmpfiles.rules = [
+      "d '${cfg.stateDir}' 0750 ${cfg.user} ${cfg.group} - -"
+
+      "L+ /etc/peroxide - - - - ${cfg.stateDir}"
+    ];
+
+    systemd.services.peroxide = {
+      requires = [ "network.target" ];
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Type = "simple";
+        User = cfg.user;
+        Group = cfg.group;
+        ExecStart = "${pkgs.peroxide}/bin/peroxide -log-file=${cfg.stateDir}/peroxide.log";
+        ExecReload = "${pkgs.coreutils}/bin/kill -SIGHUP $MAINPID";
+      };
+
+      preStart = (optionalString cfg.selfSignedCert.enable ''
+        # Create a self-signed certificate if no certificate exists.
+        if ! test -e ${cfg.stateDir}/key.pem && ! test -e ${cfg.stateDir}/cert.pem; then
+          (cd ${cfg.stateDir} && \
+            ${pkgs.peroxide}/bin/peroxide-cfg -action gen-x509 \
+              -x509-org ${cfg.selfSignedCert.organization} \
+              -x509-cn ${cfg.selfSignedCert.hostname})
+        fi
+        chmod 400 ${cfg.stateDir}/{key,cert}.pem
+      '');
+    };
+
+    environment.etc."peroxide.conf".source =
+      settingsFormat.generate "peroxide-config.yaml" cfg.settings;
+
+    users.users = optionalAttrs (cfg.user == "peroxide") {
+      peroxide = {
+        group = cfg.group;
+        home = cfg.stateDir;
+        uid = config.ids.uids.peroxide;
+      };
+    };
+
+    users.groups = optionalAttrs (cfg.group == "peroxide") {
+      peroxide.gid = config.ids.gids.peroxide;
+    };
+
+    assertions = [
+      {
+        assertion = cfg.selfSignedCert.enable ->
+          (cfg.selfSignedCert.organization != null) &&
+            (cfg.selfSignedCert.hostname != null);
+        message = "You must specify Organization and CN when using a self-signed certificate.";
```

Comment:
Do these even matter? We could probably default them to nixos-user or something.

Reply:
They do have default values, so this assert is a bit overkill.